odoo/models.py

```diff
@@ -2072,16 +2069,87 @@ def _read_group_fill_temporal(self, data, groupby, aggregated_fields, annotated_
         if first_a_gby['type'] not in ('date', 'datetime'):
             return data
         interval = first_a_gby['interval']
+        granularity = first_a_gby['granularity']
         groupby_name = groupby[0]
 
         # existing non null datetimes
-        existing = [d[groupby_name] for d in data if d[groupby_name]]
-
-        if len(existing) < 2:
+        existing = sorted([d[groupby_name] for d in data if d[groupby_name]])
+        existing_from = existing[0] if existing else False
+        existing_to = existing[-1] if existing else False
+
+        # fill temporal between bounds
+        # the context keys : fill_temporal_from and fill_temporal_to offer a way to guarantee
+        # that at least a certain date range will be returned as groups. They are also the
+        # bounds to fill. Any group outside those bounds will not be removed, but there will
+        # be no empty group added. If any of those keys is omitted, existing bounds are used,
+        # if applicable. If fill_temporal_from is not specified, and there is no data, no group
+        # is returned.
+        # edge case example:
+        # existing:           [Jan, Sep]
+        # fill_temporal_from: [Apr]
+        # fill_temporal_to:   [Jul]
+        # result:             [Jan, |Apr, May, Jun, Jul|, Sep]
+        #
+        # general usage examples:
+        # 1. We have a Forecast in a Kanban view. We are in April and we want the
+        # forecast to cover the [Apr - Jul] period.
+        #     - domain:             [(>= Apr), (<= Jul)] (limit the read_group to the forecast)
+        #     - fill_temporal_from: Apr (current month)
+        #     - fill_temporal_to:   Jul
+        # The result would be [Apr, May, Jun, Jul]
+        #
+        # 2. We want the same Forecast [Apr - Jul], but there is a chance that some
+        # records were forgotten in Jan, and we want to be able to reschedule
+        # them, but we don't need the fill_temporal to take effect from Jan.
+        #     - domain:             [(<= Jul)] (we removed the first bound)
+        #     - fill_temporal_from: Apr (current month)
+        #     - fill_temporal_to:   Jul
+        # The result would be [Jan, |Apr, May, Jun, Jul|]. We have the forecast
+        # with the fill_temporal, and every month preceding the forecast as long
+        # as there is at least one record in it (no empty month before the
+        # forecast).
+        tz = next((date.tzinfo for date in existing if date.tzinfo), False)
+        fill_from = self.env.context.get('fill_temporal_from')
+        fill_to = self.env.context.get('fill_temporal_to')
+        if fill_from:
+            fill_from = date_utils.start_of(datetime.datetime.strptime(fill_from, "%Y-%m-%d"), granularity)
+            if tz:
+                fill_from = tz.localize(fill_from)
+        elif existing_from:
+            fill_from = existing_from
+        else:
             return data
```

Comment:
That seems like an odd asymmetry between `fill_temporal_from` and `fill_temporal_to`: if only `_from` is provided then that month will be filled, but if only `_to` is provided then it won't. Why not have both bounds work the same?